Core/Rewriters/LastCallRewriter.cs

```diff
@@ -0,0 +1,168 @@
+    {
+      return node.Statements
+          .Where (s => s.IsKind (SyntaxKind.ExpressionStatement))
+          .Select (s => (ExpressionStatementSyntax) s)
+          .Select ((syntaxNode, index) => new { node = syntaxNode, index })
+          .Where (
+              s => Model!.GetSymbolInfo (node.GetOriginalNode (s.node)!.Expression).Symbol is { } symbol
+                   && rhinoMocksLastCallSymbol.Equals (symbol.OriginalDefinition.ContainingSymbol, SymbolEqualityComparer.Default))
+          .Select (s => (s.node, s.index));
+    }
+
+    private ExpressionStatementSyntax? GetLastCalledMockAsExpressionStatement (
+        BlockSyntax node,
+        (ExpressionStatementSyntax node, int index) lastCallExpressionStatement,
+        INamedTypeSymbol rhinoMocksMockRepositorySymbol)
+    {
+      var possibleLastCalledMockSymbols = GetPossibleLastCalledMocks (node, rhinoMocksMockRepositorySymbol);
+      var lastCalledExpressions = node!.Statements.Where (s => s is ExpressionStatementSyntax { Expression: InvocationExpressionSyntax }).ToList();
+
+      return lastCalledExpressions
+          .Select ((syntaxNode, index) => new { node = syntaxNode, index })
+          .LastOrDefault (
+              e => Model.GetSymbolInfo (node.GetOriginalNode (e.node)!.GetFirstIdentifierName()).Symbol is { } symbol
+                   && possibleLastCalledMockSymbols.Contains (symbol)
+                   && e.index < lastCallExpressionStatement.index)?.node as ExpressionStatementSyntax;
+    }
+
+    private IEnumerable<ISymbol?> GetPossibleLastCalledMocks (BlockSyntax node, INamedTypeSymbol rhinoMocksMockRepositorySymbol)
+    {
+      return GetAllMockedFieldSymbols (node, rhinoMocksMockRepositorySymbol)
+          .Concat (GetAllLocalMockSymbolsInAncestorsAndSelf (node, rhinoMocksMockRepositorySymbol));
+    }
+
+    private IEnumerable<ISymbol?> GetAllLocalMockSymbolsInAncestorsAndSelf (BlockSyntax node, INamedTypeSymbol rhinoMocksMockRepositorySymbol)
+    {
+      return node.GetOriginalNode (node.Statements.First())!.AncestorsAndSelf()
+          .Where (s => s.IsKind (SyntaxKind.Block))
+          .Select (s => (BlockSyntax) s)
+          .SelectMany (s => GetAllLocalMockSymbols (s, rhinoMocksMockRepositorySymbol));
+    }
+
+    private IEnumerable<ISymbol?> GetAllLocalMockSymbols (BlockSyntax node, INamedTypeSymbol rhinoMocksMockRepositorySymbol)
+    {
+      return node.Statements
+          .Where (s => s.IsKind (SyntaxKind.LocalDeclarationStatement))
+          .Select (s => (LocalDeclarationStatementSyntax) s)
+          .Select (s => s.Declaration.Variables)
+          .Where (s => IsRhinoMocksMock (s, rhinoMocksMockRepositorySymbol))
+          .SelectMany (s => s)
+          .Select (s => Model.GetDeclaredSymbol (s));
+    }
+
+    private IEnumerable<ISymbol?> GetAllMockedFieldSymbols (BlockSyntax node, INamedTypeSymbol rhinoMocksMockRepositorySymbol)
```

Comment:
Wouldn't it be easier to just get the symbol of the current type & filter the members for mock types? That should be far more readable than filtering the syntax for field declarations + fetching the associated symbols.